### PR TITLE
fix: fast timers and event loop lag

### PR DESCRIPTION
**Pull request by ronag**
When there is significant event loop lag timers can prematurely expire.

Fixes: https://github.com/nodejs/undici/issues/1976

### CODE REVIEW

**Comment by ronag** [comment]
@kajiiiro would you like helping adding a test?

**Review by mcollina — COMMENTED** [review]
Can you add a test?

File: lib/timers.js [hunk 1]
```diff
@@ -13,13 +13,15 @@ function onTimeout () {
   while (idx < len) {
     const timer = fastTimers[idx]
 
-    if (timer.expires && fastNow >= timer.expires) {
-      timer.expires = 0
+    if (timer.state === 0) {
+      timer.state = fastNow + timer.delay
+    } else if (timer.state > 0 && fastNow >= timer.state) {
```
Comment (kajiiiro) [inline lib/timers.js:18]: @ronag 
expired check is executed in next event loop(after 1000msec).
it is executed in same event loop?
```suggestion
    }
    if (timer.state > 0 && fastNow >= timer.state) {
```

**Comment by kajiiiro** [comment]
is this stopped?

> would you like helping adding a test?

Sorry, I have no idea. 
the test should confirm 'request timeout is not fired shorter than delay'
but it is hard...

- event loop interval is about 1sec
- built-in setimeout is used under 1sec

**Review by kajiiiro — APPROVED** [review]

**Comment by ronag** [comment]
I'll work on a test when I have time.

**Comment by ronag** [comment]
@mcollina looking into this, I realize that making a test for this is quite tricky and might require workers, etc... I suggest we either land this without a test or skip it. The case it is trying to fix is when you have significant event loop lag, which is an edge case...

**Review by mcollina — APPROVED** [review]
lgtm

(I trust you, but I would prefer to have a test anyway).

Let's land this and create an issue about writing this test?
